File: peacedesign/src/main/java/com/peacedesign/android/widget/sheet/PeaceBottomSheet.java
package com.peacedesign.android.widget.sheet;

import static android.view.ViewGroup.LayoutParams.MATCH_PARENT;
import static android.view.ViewGroup.LayoutParams.WRAP_CONTENT;

import android.annotation.SuppressLint;
import android.app.Dialog;
import android.content.Context;
import android.content.DialogInterface;
import android.graphics.drawable.Drawable;
import android.os.Bundle;
import android.text.TextUtils;
import android.view.ContextThemeWrapper;
import android.view.Gravity;
import android.view.KeyEvent;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.view.Window;
import android.view.inputmethod.InputMethodManager;
import android.widget.LinearLayout;
import android.widget.LinearLayout.LayoutParams;

import androidx.annotation.FloatRange;
import androidx.annotation.LayoutRes;
import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.annotation.StringRes;
import androidx.appcompat.widget.AppCompatImageView;
import androidx.appcompat.widget.AppCompatTextView;
import androidx.core.content.ContextCompat;
import androidx.fragment.app.DialogFragment;
import androidx.fragment.app.FragmentManager;

import com.google.android.material.bottomsheet.BottomSheetBehavior;
import com.google.android.material.bottomsheet.BottomSheetBehavior.State;
import com.google.android.material.bottomsheet.BottomSheetDialog;
import com.google.android.material.bottomsheet.BottomSheetDialogFragment;
import com.peacedesign.R;
import com.peacedesign.android.utils.Dimen;
import com.peacedesign.android.utils.DrawableUtils;
import com.peacedesign.android.utils.ResUtils;
import com.peacedesign.android.utils.ViewUtils;
import com.peacedesign.android.utils.WindowUtils;

import java.io.Serializable;

public class PeaceBottomSheet extends BottomSheetDialogFragment {
    private PeaceBottomSheetParams P;
    private OnPeaceBottomSheetShowListener mOnPeaceBottomSheetShowListener;
    private OnPeaceBottomSheetDismissListener mOnPeaceBottomSheetDismissListener;
    private LinearLayout mDialogLayout;

    public PeaceBottomSheet() {
        P = new PeaceBottomSheetParams();
    }

    @Override
    public void onSaveInstanceState(@NonNull Bundle outState) {
        super.onSaveInstanceState(outState);
        outState.putSerializable("params", P);
    }

    @Override
    public void onCreate(@Nullable Bundle savedInstanceState) {
        setStyle(DialogFragment.STYLE_NORMAL, R.style.PeaceBottomSheetTheme);
        if (savedInstanceState != null) {
            P = (PeaceBottomSheetParams) savedInstanceState.getSerializable("params");
        }

        super.onCreate(savedInstanceState);
        if (P.sheetBGColor == -1) {
            P.sheetBGColor = ContextCompat.getColor(getContext(), R.color.colorBackgroundSheetDialog);
        }
    }

    @NonNull
    @Override
    public PeaceBottomSheetDialog onCreateDialog(@Nullable Bundle savedInstanceState) {
        return new PeaceBottomSheetDialog(getContext(), getTheme(), P);
    }

    protected View prepareDialogLayout(Context context, PeaceBottomSheetParams params) {
        mDialogLayout = new LinearLayout(context);
        mDialogLayout.setOrientation(LinearLayout.VERTICAL);

        setupHeader(mDialogLayout, params);
        setupContentView(mDialogLayout, params);

        return mDialogLayout;
    }

    private void resolveTitle() {
        if (P.headerTitle == null) {
            P.headerTitle = getContext().getString(P.headerTitleResource);
        }
    }

    protected void setupHeader(ViewGroup dialogLayout, PeaceBottomSheetParams params) {
        if (!P.headerShown) {
            return;
        }

        resolveTitle();
        final boolean hasTitle = !TextUtils.isEmpty(params.headerTitle);
        if (!hasTitle && params.disableDragging) {
            return;
        }

        LinearLayout headerView = createHeaderView(dialogLayout);

        prepareDragIcon(headerView, params);
        prepareTitleView(headerView, params, false);

        if (hasTitle) {
            headerView.setBackground(getHeaderBG(dialogLayout.getContext()));
        }
    }

    private Drawable getHeaderBG(Context context) {
        return ContextCompat.getDrawable(context, R.drawable.dr_bg_sheet_dialog_header);
    }

    protected void prepareDragIcon(LinearLayout container, PeaceBottomSheetParams params) {
        if (params.disableDragging) {
            return;
        }
        AppCompatImageView dragIcon = new AppCompatImageView(container.getContext());
        dragIcon.setId(R.id.dragIcon);
        dragIcon.setImageResource(R.drawable.dr_icon_drag);

        LayoutParams lp = new LayoutParams(WRAP_CONTENT, WRAP_CONTENT);
        lp.topMargin = Dimen.dp2px(container.getContext(), 10);
        container.addView(dragIcon, 0, lp);
    }

    protected void prepareTitleView(@NonNull LinearLayout container, PeaceBottomSheetParams params, boolean isUpdating) {
        boolean hasTitle = !TextUtils.isEmpty(params.headerTitle);
        if (!isUpdating && !hasTitle) {
            return;
        }

        AppCompatTextView titleView = container.findViewById(R.id.title);
        if (hasTitle && titleView == null) {
            titleView = new AppCompatTextView(new ContextThemeWrapper(container.getContext(), resolveTitleThemeId()));
            titleView.setId(R.id.title);
            container.addView(titleView);
        }

        if (titleView != null) {
            titleView.setText(params.headerTitle);
            titleView.setVisibility(hasTitle ? View.VISIBLE : View.GONE);
        }
    }

    public void updateHeaderTitle() {
        if (mDialogLayout == null) {
            return;
        }

        resolveTitle();
        final boolean hasTitle = !TextUtils.isEmpty(P.headerTitle);

        LinearLayout headerView = mDialogLayout.findViewById(R.id.peaceBottomSheetHeaderView);
        if (hasTitle && headerView == null) {
            headerView = createHeaderView(mDialogLayout);
        }

        if (headerView != null) {
            prepareTitleView(headerView, P, true);
            headerView.setBackground(hasTitle ? getHeaderBG(mDialogLayout.getContext()) : null);
        }
    }

    private LinearLayout createHeaderView(ViewGroup dialogLayout) {
        LinearLayout headerView = new LinearLayout(dialogLayout.getContext());
        headerView.setId(R.id.peaceBottomSheetHeaderView);
        headerView.setOrientation(LinearLayout.VERTICAL);
        headerView.setGravity(Gravity.CENTER);
        dialogLayout.addView(headerView, 0);
        return headerView;
    }

    protected int resolveTitleThemeId() {
        return R.style.PeaceBottomSheetTitleStyle;
    }

    protected void setupContentView(@NonNull LinearLayout dialogLayout, PeaceBottomSheetParams params) {
        if (params.mContentView == null) {
            if (params.mContentViewResId != 0) {
                params.mContentView = LayoutInflater.from(getContext()).inflate(params.mContentViewResId, dialogLayout, false);
            }
        }

        if (params.mContentView != null) {
            ViewUtils.removeView(params.mContentView);
            dialogLayout.addView(params.mContentView);
        }

        if (!params.headerShown && params.mContentView != null) {
            View closeBtn = params.mContentView.findViewById(R.id.close);
            if (closeBtn != null) {
                closeBtn.setOnClickListener(v -> dismiss());
            }
        }
    }

    @SuppressLint("RestrictedApi")
    @Override
    public void setupDialog(@NonNull Dialog dialog, int style) {
        super.setupDialog(dialog, style);
        setupDialogInternal(dialog, style, P);
    }

    protected void setupDialogInternal(Dialog dialog, int style, PeaceBottomSheetParams params) {
        View dialogLayout = prepareDialogLayout(getContext(), params);
        dialog.setContentView(dialogLayout);

        setupDialogStyles(dialog, dialogLayout, params);
    }

    protected void setupDialogStyles(Dialog dialog, View dialogLayout, PeaceBottomSheetParams P) {
        Window window = dialog.getWindow();
        window.getDecorView().setClipToOutline(true);

        dialogLayout.setClipToOutline(true);
        ((ViewGroup) dialogLayout).setClipChildren(true);

        ViewGroup dialogModal = (ViewGroup) dialogLayout.getParent();
        dialogModal.setClipToOutline(true);
        dialogModal.setClipChildren(true);

        setupFullHeight(dialogModal);
        if (!P.supportsRoundedCorners) {
            setupModalBackground(dialogModal, false);
        }

        BottomSheetBehavior<View> bottomSheetBehavior = BottomSheetBehavior.from(dialogModal);
        bottomSheetBehavior.setHideable(P.cancellable && !P.hideOnSwipe);
        bottomSheetBehavior.setDraggable(P.cancellable && !P.disableDragging);
        bottomSheetBehavior.setState(P.initialBehaviorState);
        if (WindowUtils.isLandscapeMode(getContext())) {
            bottomSheetBehavior.setSkipCollapsed(true);
        }
        bottomSheetBehavior.addBottomSheetCallback(new BottomSheetBehavior.BottomSheetCallback() {
            @Override
            public void onStateChanged(@NonNull View bottomSheet, int newState) {
                setupDialogOnStateChange(dialog, dialogModal, newState);

                View focus = dialog.getCurrentFocus();
                if (focus != null) {
                    InputMethodManager imm = (InputMethodManager) getContext().getSystemService(Context.INPUT_METHOD_SERVICE);
                    imm.hideSoftInputFromWindow(focus.getWindowToken(), 0);
                }
            }

            @Override
            public void onSlide(@NonNull View bottomSheet, float slideOffset) {
                /*if (slideOffset <= 0) {
                    float mult = 1 + slideOffset;
                    window.setDimAmount(P.windowDimAmount * mult);
                }*/
            }
        });

        dialog.setOnKeyListener((dialogInterface, keyCode, event) -> onKey((BottomSheetDialog) dialogInterface, keyCode, event));
        dialog.setOnShowListener(dialogInterface -> {
            if (mOnPeaceBottomSheetShowListener != null) mOnPeaceBottomSheetShowListener.onShow();
            setupDialogOnStateChange(dialog, dialogModal, P.initialBehaviorState);
        });

        dialogModal.addOnLayoutChangeListener((v, left, top, right, bottom, oldLeft, oldTop, oldRight, oldBottom) -> {
            setupDialogOnStateChange(dialog, dialogModal, bottomSheetBehavior.getState());
        });
    }

    protected boolean onKey(@NonNull BottomSheetDialog dialog, int keyCode, @NonNull KeyEvent event) {
        return false;
    }

    private void setupFullHeight(View modal) {
        if (P.fullHeight) {
            ViewGroup.LayoutParams params = modal.getLayoutParams();
            params.height = MATCH_PARENT;
            modal.setLayoutParams(params);
        }
    }

    private void setupModalBackground(View modal, boolean isOnFullHeight) {
        float[] radii = null;

        if (P.supportsRoundedCorners) {
            boolean cornerFlag = P.resetRoundedCornersOnFullHeight && isOnFullHeight;
            if (!cornerFlag) {
                radii = Dimen.createRadiiForBGInDP(getContext(), 15, 15, 0, 0);
            }
        }

        Drawable background = DrawableUtils.createBackground(P.sheetBGColor, radii);
        modal.setBackground(background);
    }

    private void setupDialogOnStateChange(Dialog dialog, View dialogModal, int newState) {
        boolean isExpanded = newState == BottomSheetBehavior.STATE_EXPANDED;
        boolean isHeightFilled = dialogModal.getHeight() >= (Dimen.getWindowHeight(getContext()) - 10);
        boolean isOnFullHeight = isExpanded && isHeightFilled;

        setupModalBackground(dialogModal, isOnFullHeight);

        if (!P.resetRoundedCornersOnFullHeight) {
            return;
        }

        Window window = dialog.getWindow();
        window.setDimAmount(isOnFullHeight ? 0 : P.windowDimAmount);
        window.setStatusBarColor(isOnFullHeight ? P.sheetBGColor : 0);

        if (!WindowUtils.isNightMode(getContext())) {
            if (isOnFullHeight) {
                WindowUtils.setLightStatusBar(window);
            } else {
                WindowUtils.clearLightStatusBar(window);
            }
        }
    }

    public final void show(@NonNull FragmentManager fragmentManager) {
        show(fragmentManager, getClass().getSimpleName());
    }

    public void setContentView(@NonNull View contentView) {
        P.mContentView = contentView;
    }

    public void setContentView(@LayoutRes int layoutId) {
        P.mContentViewResId = layoutId;
    }

    public void setHeaderTitle(CharSequence title) {
        P.headerTitle = title;
    }

    public void setHeaderTitle(@StringRes int titleRes) {
        P.headerTitleResource = titleRes;
    }

    @Override
    public void setCancelable(boolean cancelable) {
        super.setCancelable(cancelable);
        P.cancellable = cancelable;
    }

    public void disableDragging(boolean flag) {
        P.disableDragging = flag;
    }

    public void setHideOnSwipe(boolean flag) {
        P.hideOnSwipe = flag;
    }

    public void disableOutsideTouch(boolean flag) {
        P.disableOutsideTouch = flag;
    }

    public void disableBackKey(boolean flag) {
        P.disableBackKey = flag;
    }

    public void setInitialBehaviorState(@State int state) {
        P.initialBehaviorState = state;
    }

    public void setFullHeight(boolean flag) {
        P.fullHeight = flag;
    }

    public void setHeaderShown(boolean shown) {
        P.headerShown = shown;
    }

    @NonNull
    public PeaceBottomSheetParams getDialogParams() {
        return P;
    }

    public void applyParams(@NonNull PeaceBottomSheetParams params) {
        P = params;
    }

    public void setOnShowListener(@NonNull OnPeaceBottomSheetShowListener listener) {
        mOnPeaceBottomSheetShowListener = listener;
    }

    public void setOnDismissListener(@NonNull OnPeaceBottomSheetDismissListener listener) {
        mOnPeaceBottomSheetDismissListener = listener;
    }

    public LinearLayout getDialogLayout() {
        return mDialogLayout;
    }

    @NonNull
    @Override
    public Context getContext() {
        //noinspection ConstantConditions
        return super.getContext();
    }

    public int getSheetBGColor() {
        return P.sheetBGColor;
    }

    @Override
    public void onDismiss(@NonNull DialogInterface dialog) {
        super.onDismiss(dialog);
        if (mOnPeaceBottomSheetDismissListener != null) mOnPeaceBottomSheetDismissListener.onDismissed();
    }

    public boolean isShowing() {
        return isAdded();
    }

    @Override
    public void dismiss() {
        try {
            super.dismiss();
        } catch (Exception ignored) {}
    }

    public interface OnPeaceBottomSheetShowListener {
        void onShow();
    }

    public interface OnPeaceBottomSheetDismissListener {
        void onDismissed();
    }

    public static class PeaceBottomSheetParams implements Serializable {
        public boolean disableDragging;
        public boolean hideOnSwipe;
        public boolean disableOutsideTouch;
        public boolean cancellable = true;
        public boolean disableBackKey;
        public boolean supportsRoundedCorners = true;
        public boolean resetRoundedCornersOnFullHeight = true;
        public boolean supportsOverlayBackground = true;
        public boolean supportsAnimations = true;
        public boolean supportsEnterAnimation = true;
        public boolean supportsExitAnimation = true;
        public boolean fullHeight;
        public boolean headerShown = true;
        public int sheetBGColor = -1;
        @FloatRange(from = 0f, to = 1f)
        public float windowDimAmount = 0.6f;
        @State
        public int initialBehaviorState = BottomSheetBehavior.STATE_EXPANDED;
        @Nullable
        public CharSequence headerTitle;
        @StringRes
        public int headerTitleResource;
        @Nullable
        protected transient View mContentView;
        @LayoutRes
        protected int mContentViewResId;

        public boolean supportsNoAnimation() {
            return !supportsAnimations || (!supportsEnterAnimation && !supportsExitAnimation);
        }

        public boolean supportsEnterAnimationOnly() {
            return supportsAnimations && supportsEnterAnimation && !supportsExitAnimation;
        }

        public boolean supportsExitAnimationOnly() {
            return supportsAnimations && !supportsEnterAnimation && supportsExitAnimation;
        }


        public void setContentView(@LayoutRes int viewResId) {
            mContentViewResId = viewResId;
        }

        public void setContentView(@NonNull View contentView) {
            mContentView = contentView;
        }
    }
}

File: app/src/main/java/com/quranapp/android/widgets/bottomSheet/PeaceBottomSheet.kt
package com.quranapp.android.widgets.bottomSheet

import android.annotation.SuppressLint
import android.app.Dialog
import android.content.Context
import android.content.DialogInterface
import android.graphics.drawable.Drawable
import android.os.Bundle
import android.text.TextUtils
import android.view.*
import android.view.ViewGroup.LayoutParams.WRAP_CONTENT
import android.view.inputmethod.InputMethodManager
import android.widget.LinearLayout
import androidx.appcompat.widget.AppCompatImageView
import androidx.appcompat.widget.AppCompatTextView
import androidx.core.content.ContextCompat
import androidx.fragment.app.FragmentManager
import com.google.android.material.bottomsheet.BottomSheetBehavior
import com.google.android.material.bottomsheet.BottomSheetDialog
import com.google.android.material.bottomsheet.BottomSheetDialogFragment
import com.peacedesign.android.utils.Dimen
import com.peacedesign.android.utils.DrawableUtils
import com.peacedesign.android.utils.WindowUtils
import com.quranapp.android.R
import com.quranapp.android.utils.extensions.*

open class PeaceBottomSheet : BottomSheetDialogFragment() {
    var params = PeaceBottomSheetParams()
    var onPeaceBottomSheetShowListener: OnPeaceBottomSheetShowListener? = null
    var onDismissListener: OnPeaceBottomSheetDismissListener? = null

    private var dialogLayout: LinearLayout? = null

    override fun onSaveInstanceState(outState: Bundle) {
        super.onSaveInstanceState(outState)
        outState.putSerializable("params", params)
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        setStyle(STYLE_NORMAL, R.style.PeaceBottomSheetTheme)

        if (savedInstanceState != null && savedInstanceState.containsKey("params")) {
            params = savedInstanceState.serializableExtra("params")!!
        }

        super.onCreate(savedInstanceState)

        if (params.sheetBGColor == -1) {
            params.sheetBGColor = requireContext().color(R.color.colorBackgroundSheetDialog)
        }
    }

    override fun onCreateDialog(savedInstanceState: Bundle?): PeaceBottomSheetDialog {
        return PeaceBottomSheetDialog(requireContext(), theme, params)
    }

    protected open fun prepareDialogLayout(context: Context, params: PeaceBottomSheetParams): View {
        dialogLayout = LinearLayout(context).apply {
            orientation = LinearLayout.VERTICAL
        }

        setupHeader(dialogLayout!!, params)
        setupContentView(dialogLayout!!, params)

        return dialogLayout!!
    }

    private fun resolveTitle(context: Context) {
        if (params.headerTitleResource == 0) return

        params.headerTitle = params.headerTitle ?: context.getString(params.headerTitleResource)
    }

    protected open fun setupHeader(dialogLayout: ViewGroup, params: PeaceBottomSheetParams) {
        if (!this.params.headerShown) return

        resolveTitle(dialogLayout.context)

        val hasTitle = !params.headerTitle.isNullOrEmpty()
        if (!hasTitle && params.disableDragging) return

        val headerView = createHeaderView(dialogLayout)

        prepareDragIcon(headerView, params)
        prepareTitleView(headerView, params, false)

        if (hasTitle) {
            headerView.background = getHeaderBG(dialogLayout.context)
        }
    }

    private fun getHeaderBG(context: Context): Drawable {
        return context.drawable(R.drawable.dr_bg_sheet_dialog_header)
    }

    protected open fun prepareDragIcon(container: LinearLayout, params: PeaceBottomSheetParams) {
        if (params.disableDragging) return

        val dragIcon = AppCompatImageView(container.context).apply {
            id = R.id.dragIcon
            setImageResource(R.drawable.dr_icon_drag)
        }

        container.addView(
            dragIcon,
            0,
            LinearLayout.LayoutParams(WRAP_CONTENT, WRAP_CONTENT).apply {
                topMargin = Dimen.dp2px(container.context, 10f)
            }
        )
    }

    protected open fun prepareTitleView(
        container: LinearLayout,
        params: PeaceBottomSheetParams,
        isUpdating: Boolean
    ) {
        val hasTitle = !TextUtils.isEmpty(params.headerTitle)
        if (!isUpdating && !hasTitle) return

        var titleView: AppCompatTextView? = container.findViewById(R.id.title)

        if (hasTitle && titleView == null) {
            titleView = AppCompatTextView(
                ContextThemeWrapper(container.context, resolveTitleThemeId())
            )
            titleView.id = R.id.title
            container.addView(titleView)
        }

        titleView?.let {
            it.text = params.headerTitle
            it.visibility = if (hasTitle) View.VISIBLE else View.GONE
        }
    }

    fun updateHeaderTitle() {
        val layout = dialogLayout ?: return

        resolveTitle(dialogLayout!!.context)

        val hasTitle = !params.headerTitle.isNullOrEmpty()

        var headerView: LinearLayout? = layout.findViewById(R.id.peaceBottomSheetHeaderView)

        if (hasTitle && headerView == null) {
            headerView = createHeaderView(layout)
        }

        headerView?.let {
            prepareTitleView(it, params, true)
            it.background = if (hasTitle) getHeaderBG(layout.context) else null
        }
    }

    private fun createHeaderView(dialogLayout: ViewGroup): LinearLayout {
        val headerView = LinearLayout(dialogLayout.context).apply {
            id = R.id.peaceBottomSheetHeaderView
            orientation = LinearLayout.VERTICAL
            gravity = Gravity.CENTER
        }

        dialogLayout.addView(headerView, 0)

        return headerView
    }

    protected open fun resolveTitleThemeId(): Int {
        return R.style.PeaceBottomSheetTitleStyle
    }

    protected open fun setupContentView(dialogLayout: LinearLayout, params: PeaceBottomSheetParams) {
        if (params.contentView == null) {
            if (params.contentViewResId != 0) {
                params.contentView = LayoutInflater.from(context).inflate(
                    params.contentViewResId,
                    dialogLayout,
                    false
                )
            }
        }

        if (params.contentView != null) {
            params.contentView.removeView()
            dialogLayout.addView(params.contentView)
        }

        if (!params.headerShown && params.contentView != null) {
            val closeBtn: View? = params.contentView!!.findViewById(R.id.close)
            closeBtn?.setOnClickListener { dismiss() }
        }
    }

    @SuppressLint("RestrictedApi")
    override fun setupDialog(dialog: Dialog, style: Int) {
        super.setupDialog(dialog, style)
        setupDialogInternal(dialog, style, params)
    }

    protected open fun setupDialogInternal(
        dialog: Dialog,
        style: Int,
        params: PeaceBottomSheetParams
    ) {
        val dialogLayout = prepareDialogLayout(dialog.context, params)
        dialog.setContentView(dialogLayout)
        setupDialogStyles(dialog, dialogLayout, params)
    }

    protected open fun setupDialogStyles(
        dialog: Dialog,
        dialogLayout: View,
        P: PeaceBottomSheetParams
    ) {
        dialog.window?.decorView?.clipToOutline = true

        dialogLayout.clipToOutline = true
        (dialogLayout as ViewGroup).clipChildren = true

        val dialogModal = (dialogLayout.getParent() as ViewGroup).apply {
            clipToOutline = true
            clipChildren = true
        }

        setupFullHeight(dialogModal)
        if (!P.supportsRoundedCorners) {
            setupModalBackground(dialogModal, false)
        }

        val bottomSheetBehavior = BottomSheetBehavior.from<View>(dialogModal).apply {
            isHideable = P.cancellable && !P.hideOnSwipe
            isDraggable = P.cancellable && !P.disableDragging

            state = P.initialBehaviorState

            if (WindowUtils.isLandscapeMode(dialogLayout.context)) {
                skipCollapsed = true
            }
        }

        bottomSheetBehavior.addBottomSheetCallback(object : BottomSheetBehavior.BottomSheetCallback() {
            override fun onStateChanged(bottomSheet: View, newState: Int) {
                setupDialogOnStateChange(dialog, dialogModal, newState)

                dialog.currentFocus?.let {
                    ContextCompat.getSystemService(dialog.context, InputMethodManager::class.java)
                        ?.hideSoftInputFromWindow(it.windowToken, 0)
                }
            }

            override fun onSlide(bottomSheet: View, slideOffset: Float) {
                /*if (slideOffset <= 0) {
                    float mult = 1 + slideOffset;
                    window.setDimAmount(P.windowDimAmount * mult);
                }*/
            }
        })

        dialog.setOnKeyListener { dialogInterface, keyCode, event ->
            onKey(
                dialogInterface as BottomSheetDialog,
                keyCode,
                event
            )
        }

        dialog.setOnShowListener {
            onPeaceBottomSheetShowListener?.onShow()
            setupDialogOnStateChange(dialog, dialogModal, P.initialBehaviorState)
        }

        dialogModal.addOnLayoutChangeListener { _, _, _, _, _, _, _, _, _ ->
            setupDialogOnStateChange(
                dialog,
                dialogModal,
                bottomSheetBehavior.state
            )
        }
    }

    protected open fun onKey(dialog: BottomSheetDialog, keyCode: Int, event: KeyEvent): Boolean {
        return false
    }

    private fun setupFullHeight(modal: View) {
        if (params.fullHeight) {
            modal.layoutParams.height = ViewGroup.LayoutParams.MATCH_PARENT
            modal.requestLayout()
        }
    }

    private fun setupModalBackground(modal: View, isOnFullHeight: Boolean) {
        var radii: FloatArray? = null

        if (params.supportsRoundedCorners && !(params.resetRoundedCornersOnFullHeight && isOnFullHeight)) {
            radii = Dimen.createRadiiForBGInDP(context, 15f, 15f, 0f, 0f)
        }

        modal.background = DrawableUtils.createBackground(params.sheetBGColor, radii)
    }

    private fun setupDialogOnStateChange(dialog: Dialog, dialogModal: View, newState: Int) {
        val isExpanded = newState == BottomSheetBehavior.STATE_EXPANDED
        val isHeightFilled = dialogModal.height >= (dialogModal.context.getWindowHeight() - 10)
        val isOnFullHeight = isExpanded && isHeightFilled

        setupModalBackground(dialogModal, isOnFullHeight)

        if (!params.resetRoundedCornersOnFullHeight) return

        val window = dialog.window?.apply {
            setDimAmount(if (isOnFullHeight) 0F else params.windowDimAmount)
            statusBarColor = if (isOnFullHeight) params.sheetBGColor else 0
        }

        if (!WindowUtils.isNightMode(dialogModal.context)) {
            if (isOnFullHeight) {
                WindowUtils.setLightStatusBar(window!!)
            } else {
                WindowUtils.clearLightStatusBar(window!!)
            }
        }
    }

    fun show(fragmentManager: FragmentManager) {
        if (isShowing()) return

        show(fragmentManager, javaClass.simpleName)
    }

    override fun setCancelable(cancelable: Boolean) {
        super.setCancelable(cancelable)
        params.cancellable = cancelable
    }

    fun getDialogLayout(): LinearLayout {
        return dialogLayout!!
    }

    override fun onDismiss(dialog: DialogInterface) {
        super.onDismiss(dialog)
        if (onDismissListener != null) onDismissListener!!.onDismissed()
    }

    fun isShowing() = isAdded

    override fun dismiss() {
        try {
            super.dismiss()
        } catch (ignored: Exception) {
        }
    }

    interface OnPeaceBottomSheetShowListener {
        fun onShow()
    }

    interface OnPeaceBottomSheetDismissListener {
        fun onDismissed()
    }
}
